### PR TITLE
chore: Refactor and Add tests for `AddToExistingLock`

### PR DESCRIPTION
**Pull request by mattverse**
sub-component of: #1838 

## What is the purpose of the change

`AddToExistingLock` had a bad design of returning array of locks, and then only using the first index out of the given array. 
There's no need to return an array here, but simply return the lock that we've added the token to. This PR changes the returning values of `AddToExistingLock` from an array to a single lock(not pointer referenced).

This PR also includes changing tests for `AddToExistingLock` to be TDD, adding some edge cases to the tests.

## Brief Changelog
- Change return values of `AddToExistingLock`
- Change tests for `AddToExistingLock` to be TDD

## Testing and Verifying

This change added tests 

## Documentation and Release Note

  - Does this pull request introduce a new feature or user-facing behavior changes? no
  - Is a relevant changelog entry added to the `Unreleased` section in `CHANGELOG.md`? no
  - How is the feature or change documented? not documented

### CODE REVIEW

File: x/lockup/keeper/lock.go
Comment (stackman27) [inline x/lockup/keeper/lock.go:56]: This has been a confusion for many PRs ([one example](https://github.com/osmosis-labs/osmosis/pull/1233#discussion_r866039858)). Can we add a simple comment on why we take the first lock?

Comment (mattverse) [reply x/lockup/keeper/lock.go:56]: addressed!

File: x/lockup/keeper/msg_server.go
Comment (mattverse) [reply x/lockup/keeper/msg_server.go:61]: Hrmm im not sure if I follow, the events below are only emitted when a new lock is creating?

Comment (stackman27) [reply x/lockup/keeper/msg_server.go:61]: I was moreso wondering whether the emmited events values like `lock.Owner` `lock.Coins` etc would be from this lock `lock, err := server.keeper.AddToExistingLock(ctx, owner, msg.Coins[0], msg.Duration)` or the other `lock, err = server.keeper.CreateLock(ctx, owner, msg.Coins, msg.Duration)`? 


I believe it has to be from the second i.e. CreateLock function but the lock value was deassigned there so all of those emit events values are coming from  AddToExistingLock function

Comment (stackman27) [reply x/lockup/keeper/msg_server.go:61]: Like can't we just replace `lock, err = server.keeper.CreateLock(ctx, owner, msg.Coins, msg.Duration)` to 
`_, err = server.keeper.CreateLock(ctx, owner, msg.Coins, msg.Duration)` ?

Comment (stackman27) [reply x/lockup/keeper/msg_server.go:61]: aaah! So `lock` has already been declared + assigned at `AddToExistingLock` but if it is empty we just assign lock to `CreateLock` output

I just got confused on assignment and declaration

Comment (mattverse) [reply x/lockup/keeper/msg_server.go:61]: No problem!

File: x/lockup/keeper/lock.go
Comment (p0mvn) [inline x/lockup/keeper/lock.go:62]:
```suggestion
		return lock, nil
```

File: x/lockup/keeper/lock.go
Comment (p0mvn) [inline x/lockup/keeper/lock.go:51]:
```suggestion
// Returns an empty period lock struct when a lock with the given condition does not exist.
```

Comment (p0mvn) [inline x/lockup/keeper/lock.go:51]: Also, what is returned when a lock does exist? Consider elaborating please

File: x/lockup/keeper/lock.go
Comment (p0mvn) [inline x/lockup/keeper/lock.go:62]: I have a concern - so we add tokens to lock by id but then return an old lock. Is that expected? 

The lock must be updated with tokens, maybe we need to return a new lock returned by ` AddTokensToLockByID(...)`

**Review by p0mvn — CHANGES_REQUESTED** [review]
Nice work. I worry that we might be returning an old lock from `AddToExistingLock`. Could you please confirm if that is the case? 

We should make assertions on the contents of the lock returned by `AddToExistingLock`

File: x/lockup/keeper/msg_server.go
Comment (AlpinYukseloglu) [inline x/lockup/keeper/msg_server.go:48]: Is the idea here that if the lock exists (ID > 0), then we can assume that it was successfully added to?

File: x/lockup/keeper/lock.go
Comment (p0mvn) [reply x/lockup/keeper/lock.go:62]: What if you try logging `updatedLock`? That's what I suspect is going to be outdated.

It seems to me that we update the db correctly so a future
```
lock1, err := suite.App.LockupKeeper.GetLockByID(suite.Ctx, updatedLock.ID)
```

should work as expected. However, the return from `AddToExistingLock` itself might be outdated

Comment (stackman27) [reply x/lockup/keeper/lock.go:62]: Here's the output, interesting thing it is logging the same thing when i try it with return value you suggested. 
```
Matt 
Lock: {1 osmo1v9jxgu3395kj6tfd95kj6tfd95kj6tfd0hptum 1s 0001-01-01 00:00:00 +0000 UTC 10stake}
UpdatedLock:  {1 osmo1v9jxgu3395kj6tfd95kj6tfd95kj6tfd0hptum 1s 0001-01-01 00:00:00 +0000 UTC 10stake}

Roman
Lock: {1 osmo1v9jxgu3395kj6tfd95kj6tfd95kj6tfd0hptum 1s 0001-01-01 00:00:00 +0000 UTC 10stake}
UpdatedLock: {1 osmo1v9jxgu3395kj6tfd95kj6tfd95kj6tfd0hptum 1s 0001-01-01 00:00:00 +0000 UTC 10stake}
```

I think a solution might be to return the lock from `TestLog, err := k.AddTokensToLockByID(ctx, lock.ID, owner, coin)` which actually outputs, the correct result: `TestLog: ID:1 owner:"osmo1v9jxgu3395kj6tfd95kj6tfd95kj6tfd0hptum" duration:<seconds:1 > end_time:<seconds:-62135596800 > coins:<denom:"stake" amount:"20" > `

Comment (p0mvn) [reply x/lockup/keeper/lock.go:62]: Yeah we need to be returning from `AddTokensToLockByID`

File: x/lockup/keeper/lock.go
Comment (mattverse) [reply x/lockup/keeper/lock.go:51]: done!

File: x/lockup/keeper/lock.go
Comment (mattverse) [reply x/lockup/keeper/lock.go:62]: good catch!

I took a more detailed look, currently changing this to simply return the lock ID, as that's whats only being used when this method is getting called. No reason to pass locks around!

File: x/lockup/keeper/lock_test.go
Comment (mattverse) [reply x/lockup/keeper/lock_test.go:412]: Changed the API so that it returns a lock ID and we use that lock ID in the test cases.  lmk what you think!

File: x/lockup/keeper/lock_test.go
Comment (p0mvn) [inline x/lockup/keeper/lock_test.go:376]:
```suggestion
```

File: x/lockup/keeper/lock_test.go
Comment (p0mvn) [inline x/lockup/keeper/lock_test.go:383]:
```suggestion
```

File: x/lockup/keeper/lock_test.go
Comment (p0mvn) [inline x/lockup/keeper/lock_test.go:390]:
```suggestion
```

File: x/lockup/keeper/lock.go
Comment (p0mvn) [inline x/lockup/keeper/lock.go:57]: Please consider adding a check for lock length not being 1, returning an error in such a case and a test.

I think it's useful to make sure that we catch errors as soon as possible. For example, if someone accidentally introduces a bug in `GetAccountLockedDurationNotUnlockingOnly`, returning no locks and there is no test for it, we would be able to catch it right here.

**Review by p0mvn — COMMENTED** [review]
New return with lock id LGTM, nice work!

On another look, thought about adding an extra check. Let me know what you think

File: x/lockup/keeper/lock.go
Comment (p0mvn) [inline x/lockup/keeper/lock.go:66]: Since there is no existing lock, why wouldn't this be an error?

File: x/lockup/keeper/msg_server.go
Comment (mattverse) [reply x/lockup/keeper/msg_server.go]: yes correct!

File: x/lockup/keeper/lock.go
Comment (mattverse) [reply x/lockup/keeper/lock.go:57]: I was also thinking about this as well, I think further investigation needs to be done before we change the logic that you've pointed out, since this was a logic that has been laying here for a long time.

I wanted to investigate more on this and confirm with certainty before we change it so that we assume that lock length is always 1. I can create an issue to track this! lmk what you think

File: x/lockup/keeper/lock.go
Comment (mattverse) [reply x/lockup/keeper/lock.go:66]: it's mainly because of how this API is getting used. 

this API is getting called in the msg server, we attempt to add to existing lock before creating a new lock with the user + duration combination.

Theefore, in this case, it's not an error that we couldn't find a lock with the user + duration combination.

A idea that pops up in my mind rn is maybe changing this API to return a boolean value as well to indicate whether it has been added to existing lock or not?

lmk what you think

File: x/lockup/keeper/lock.go
Comment (AlpinYukseloglu) [reply x/lockup/keeper/lock.go:57]: If that's the case, then wouldn't we have to come to a conclusion on whether lock length = 1 is a sound assumption before merging this?

Happy to help investigate if you get blocked/want a second pair of eyes

File: x/lockup/keeper/lock_test.go
Comment (AlpinYukseloglu) [inline x/lockup/keeper/lock_test.go]: Am I correct in assuming you removed `expectAddTokensToLockSuccess` wherever it was false since it's not used in the test logic (because we use `else`)?

If so, I have personally found that having an expected pass or fail on every test case makes them much easier to read and understand, especially since we have no explicit standard saying "if `expectSuccess` is omitted then assume it's a failure case"

So I have a preference to keep an expected pass/fail in each test case just for clarity, although I recognize this might be a question about testing standards so this is a light preference

File: x/lockup/keeper/lock_test.go
Comment (AlpinYukseloglu) [inline x/lockup/keeper/lock_test.go:402]: Out of curiosity, what's the point of clarifying that this is the "original" lock? Doesn't `AddToExistingLock` mutate it and make it non-original?

File: x/lockup/keeper/lock_test.go
Comment (AlpinYukseloglu) [inline x/lockup/keeper/lock_test.go:408]: Just to clarify, this is because `AddToExistingLock` now returns the lock ID instead of the full lock, correct?

File: x/lockup/keeper/lock.go
Comment (mattverse) [reply x/lockup/keeper/lock.go:57]: This was the exact logic gate(we had the same check before this PR) we had before changing so that part should be fine in terms of safteyness

File: x/lockup/keeper/lock_test.go
Comment (mattverse) [reply x/lockup/keeper/lock_test.go:402]: Good question! This is not a pointer reference to the lock so `AddToExistingLock` would not make this non-original

File: x/lockup/keeper/lock_test.go
Comment (mattverse) [reply x/lockup/keeper/lock_test.go:408]: correct!

File: x/lockup/keeper/lock.go
Comment (p0mvn) [reply x/lockup/keeper/lock.go:66]: How do you think the boolean would help? I might not be seeing it but I'm afraid that it would only overcomplicate the logic of the caller.

In my opinion, independent of the solution chosen, we should decouple the implementation details of this method from its caller.

I'd like to suggest returning an error here and then let the caller decide how they want to handle it. That is, the caller can check what error it is specifically and continue normally if the error is deemed acceptable.

Another idea is to implement a method `HasLock()` that we call before calling this method. At the same time, we can make this method assume that the lock already exists. If not, an error would be an expected return here.

Let me know what you decide

File: x/lockup/keeper/lock_test.go
Comment (AlpinYukseloglu) [reply x/lockup/keeper/lock_test.go:402]: Ah I see, thanks!

**Review by AlpinYukseloglu — APPROVED** [review]
All my comments/questions have been addressed – LGTM!

File: x/lockup/keeper/lock.go
Comment (mattverse) [reply x/lockup/keeper/lock.go:66]: The original intention behind my suggestion of adding a boolean field was to separate real errors and simple cases without matching locks, but I think your solution works as well!

Just added `HasLock` check!

File: x/lockup/keeper/lock.go
Comment (mattverse) [reply x/lockup/keeper/lock.go:57]: https://github.com/osmosis-labs/osmosis/issues/2142

File: x/lockup/keeper/lock.go
Comment (p0mvn) [reply x/lockup/keeper/lock.go:66]: Awesome! So do you think it makes sense to return an error now here, making this method only be successful if the lock exists?

Comment (mattverse) [reply x/lockup/keeper/lock.go:66]: Makes sense, this change makes it state breaking though right?

Comment (p0mvn) [reply x/lockup/keeper/lock.go:66]: Yes, it does. We shouldn't backport this

Comment (mattverse) [reply x/lockup/keeper/lock.go:66]: hrrmmm do you think this change is worth doing if its state breaking?

**Comment by mattverse** [comment]
agreed with you on adding error when there's no lock druing `AddToExistingLock`,changing it + adding a changlog entry and requesting a final review

File: x/lockup/keeper/lock.go
Comment (p0mvn) [inline x/lockup/keeper/lock.go:73]: Can we add a table-driven test for this method as well, please? Also, it might make sense to have it unexported since we, currently, only call it from the same package.

File: x/lockup/keeper/lock.go
Comment (p0mvn) [inline x/lockup/keeper/lock.go:52]:
```suggestion
// Returns the updated lock ID if successfully added coin, returns 0 and error when a lock with
// given condition does not exist, or if fails to add to lock.
```

File: x/lockup/keeper/lock_test.go
Comment (p0mvn) [inline x/lockup/keeper/lock_test.go:401]: What is the difference between `expectError` and `expectAddTokensToLockSuccess`? Can we use only `expectError`?

**Review by p0mvn — APPROVED** [review]
LGTM. Thanks for working out all edge cases here @mattverse

File: x/lockup/keeper/lock_test.go
Comment (mattverse) [reply x/lockup/keeper/lock_test.go:391]: good catch, the original intent was to create a separate case for lock failing and adding to lock failing, but these can be grouped to one. Changed to removing the field!

**Review by alexanderbez — APPROVED** [review]

**Pull request merged**
